File: cSharp/IMove.cs
namespace lean_pactheman_client {

    /// <summary>
    /// The base interface for communication with the server / player instance.<br/>
    /// IMove must be implemented by any ai interface.
    /// </summary>
    public interface IMove {
        /// <summary>
        /// A default implementation for ai interfaces that only use PerformMoveAsync.
        /// Other interfaces have to override this method.
        /// </summary>
        /// <param name="info">A <c>PlayerInfo</c> object containing the current state of the player.</param>
        /// <returns>
        /// An inplace tuple consisting of: <br/>
        ///     - <c>bool</c> SendMove: Wheter the player instance should send this move to the server <br/>
        ///     - <c>Velocity</c> Velocity: The Velocity to apply for the player instance
        /// </returns>
        (bool SendMove, Velocity Velocity) PerformMove(PlayerInfo info) {
            PerformMoveAsync(info);
            return (false, Velocity.Zero);
        }
# pragma warning disable 1998
        /// <summary>
        /// The asynchronous version of PerformMove. <br/>
        /// Meant to be used if the ai interface requires immediate PlayerState updates.
        /// </summary>
        /// <param name="info">A <c>PlayerInfo</c> object containing the current state of the player.</param>
        async void PerformMoveAsync(PlayerInfo info) {}
# pragma warning restore
    }
}

File: typescript/src/iMove.ts
import Velocity from "./classes/velocity";
import { PlayerInfo } from "./classes/player";

export type MoveResult = { sendMove: boolean, updateVelocity: Velocity };

/*
The base interface for communication with the server / player instance.
IMove must be implemented by any ai interface.
*/
export default abstract class IMove {
    /*
    A default implementation for ai interfaces that only uses PerformMoveAsync.
    Concrete implementations that don't require immediate feedback have to override this method.
    */
    performMove(info: PlayerInfo): MoveResult {
        this.performMoveAsync(info);
        return { sendMove: false, updateVelocity: Velocity.Zero };
    }
    /*
    Asynchronous version of PerformMove.
    Meant to be used if the ai interface requires immediate PlayerState updates.
    */
    performMoveAsync(info: PlayerInfo): void { }
}